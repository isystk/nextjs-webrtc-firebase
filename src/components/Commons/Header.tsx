import React, { FC } from 'react'
import { AppBar, Toolbar, IconButton } from '@material-ui/core'
import MenuIcon from '@material-ui/icons/Menu'
import { MenuContext, MenuContextProps } from '@/components/Home'

export const CommonHeader: FC = () => {
  const DEAULT_TITLE = process.env.APP_NAME
  return (
    <MenuContext.Consumer>
      {(value: MenuContextProps) => {
        const { isMenuOpen, setMenuOpen, roomName } = value
        return (
          <>
            <AppBar position="fixed" className="App-header">
              <Toolbar>
                <IconButton
                  color="inherit"
                  disabled={!roomName}
                  onClick={() => setMenuOpen(!isMenuOpen)}
                >
                  <MenuIcon />
                </IconButton>
                <div className="App-logo">{roomName || DEAULT_TITLE}</div>
              </Toolbar>
            </AppBar>
          </>
        )
      }}
    </MenuContext.Consumer>
  )
}

export default CommonHeader
