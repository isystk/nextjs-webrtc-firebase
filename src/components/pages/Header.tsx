import React, { Dispatch, SetStateAction, VFC } from 'react'
import { AppBar, Toolbar, IconButton } from '@material-ui/core'
import MenuIcon from '@material-ui/icons/Menu'
import { useSelector } from 'react-redux'
import { Client } from '@/store/StoreTypes'

type Props = {
  isMenuOpen: boolean
  setMenuOpen: Dispatch<SetStateAction<boolean>>
  client: Client
}

const CommonHeader: VFC<Props> = ({ isMenuOpen, setMenuOpen }) => {
  const { client } = useSelector((state: {client: Client}) => state.client)
  const DEAULT_TITLE = process.env.APP_NAME
  return (
    <>
      <AppBar position="fixed" className="App-header">
        <Toolbar>
          <IconButton
            color="inherit"
            disabled={!client.roomName}
            onClick={() => setMenuOpen(!isMenuOpen)}
          >
            <MenuIcon />
          </IconButton>
          <div className="App-logo">{client.roomName || DEAULT_TITLE}</div>
        </Toolbar>
      </AppBar>
    </>
  )
}

export default CommonHeader
