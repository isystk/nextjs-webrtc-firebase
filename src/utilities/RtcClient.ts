import { getDatabase, getAuth } from '@/utilities/firebase'
import WebRtc from '@/utilities/WebRtc'
import DisplayShare from "@/utilities/DisplayShare";

export type Self = {
  clientId?: string
  name: string
  videoOff?: boolean
  muted?: boolean
}
export type Room = {
  roomId?: string
  name: string
}
export type Member = {
  clientId: string
  name: string
  webRtc: WebRtc | null
  status: string
}
type Members = {
  [key: string]: Member
}
export type Chat = {
  isOpen: boolean
}
export type Share = {
  clientId?: string
  mediaStream: MediaStream | null
  members: ShareMembers
}
export type ShareMember = {
  clientId: string
  webRtc: DisplayShare | null
}
type ShareMembers = {
  [key: string]: ShareMember
}

export default class RtcClient {
  _setRtcClient: (rtcClient: RtcClient) => void
  constraints: { audio: boolean, video: boolean }
  mediaStream: MediaStream | null
  members: Members
  room: Room
  self: Self
  share: Share
  chat: Chat

  constructor(setRtcClient: (rtcClient: RtcClient) => void) {
    console.log('Initial RtcClient')
    this._setRtcClient = setRtcClient
    this.constraints = { audio: true, video: true }
    this.mediaStream = null
    this.members = {}
    this.room = { roomId: undefined, name: '' }
    this.self = { clientId: undefined, name: '' }
    this.share = { clientId: undefined, mediaStream: null, members: {}}
    this.chat = { isOpen: false }
  }

  async setRtcClient() {
    await this._setRtcClient(this)
  }

  // カメラの使用許可を取得する
  async getUserMedia() {
    try {
      this.mediaStream = await navigator.mediaDevices.getUserMedia(this.constraints)
    } catch (error) {
      console.error(error)
    }
  }

  // メディア(音声とビデオ)の仕様を許可する
  async setMediaStream() {
    await this.getUserMedia()
    // await this.getDisplayMedia()
    await this.setRtcClient()
  }

  async setLocalPeerName(localPeerName: string) {
    this.self.name = localPeerName
    await this.setRtcClient()
  }

  async setRoomName(roomName: string) {
    const key = getDatabase().push({
      name: roomName,
    }).key
    this.room = {
      roomId: key + '',
      name: roomName,
    }
    // TODO ここにawaitを付けると何故か動作しない
    getDatabase(this.room.roomId).update(this.room)
    await this.setRtcClient()
  }

  async setRoomId(roomId: string) {
    await getDatabase(roomId).once('value', async (snapshot) => {
      const data = snapshot.val()
      if (data === null) return
      const { roomId, name } = data
      this.room = {
        roomId,
        name,
      }
      await this.setRtcClient()
    })
  }

  // 映像のオン・オフを切り替える
  async toggleVideo() {
    this.self.videoOff = !this.self.videoOff;
    await this.setRtcClient()
    if (Object.keys(this.members).length === 0) return
    Object.keys(this.members).forEach((key) => {
      this.members[key].webRtc?.toggleVideo()
    })
  }

  // 音声のオン・オフを切り替える
  async toggleAudio() {
    this.self.muted = !this.self.muted;
    await this.setRtcClient()
    if (Object.keys(this.members).length === 0) return
    Object.keys(this.members).forEach((key) => {
      this.members[key].webRtc?.toggleAudio()
    })
  }

  // チャットの表示・非表示を切り替える
  async openChat() {
    this.chat.isOpen = true;
    await this.setRtcClient()
  }
  async closeChat() {
    this.chat.isOpen = false;
    await this.setRtcClient()
  }

  // 画面共有を開始する
  async startShare() {
    try {
      const mediaStream = await navigator.mediaDevices.getDisplayMedia({ audio: false, video: true })
      this.share = {...this.share, clientId: this.self.clientId, mediaStream}

      if (Object.keys(this.members).length === 0) return
      Object.keys(this.members).forEach((key) => {
        const member = {...this.members[key]} as ShareMember
        this.addShare(member)
      })

      // 画面共有を開始することをすべてのメンバーに通知する
      await this.databaseBroadcastRef.set({
        type: 'share',
        clientId: this.self.clientId,
      })

      Object.keys(this.share.members).forEach(async (key) => {
        const member = this.share.members[key]
        member.webRtc?.addTracks(mediaStream);
        await member.webRtc?.offer()
      })

      await this.setRtcClient()
    } catch (error) {
      console.error(error)
    }
  }

  // 画面共有を開始した時やshareを受信したら画面共有を表示する
  async addShare(data: ShareMember) {
    if (this.self.clientId && this.room.roomId) {
      data.webRtc = new DisplayShare(
          this.room.roomId,
          this.self.clientId,
          data.clientId
      )
      await data.webRtc.startListening()
    }
    const newMember = {
      [data.clientId]: data,
    }
    this.share.members = { ...this.members, ...newMember }
    await this.setRtcClient()
  }

  async signOut() {
    console.log('logout')
    await this.disconnect()
    await getAuth().signOut();
    this.self = { clientId: undefined, name: '' }
    await this.setRtcClient()
  }

  async disconnect() {
    console.log('disconnect')
    await this.databaseMembersRef(this.self.clientId).remove()
    this.room = { roomId: undefined, name: '' }
    await this.setRtcClient()
  }

  // ルームに参加する
  async join() {
    try {
      // joinを初期化する
      await this.databaseJoinRef().remove()

      // メンバーに自分を追加する
      const key = await this.databaseMembersRef().push({
        name: this.self.name,
      }).key
      this.self = {
        clientId: key + '',
        name: this.self.name,
      }
      await this.databaseMembersRef(this.self.clientId).update(this.self)

      // シグナリングサーバーをリスンする
      await this.startListening()

      // 1. Aさんがルームに入ったらブロードキャストですべてのメンバーにjoinを送信する
      console.log('send join', this.room.roomId, this.self)
      await this.databaseJoinRef(this.self.clientId).set({
        ...this.self,
        type: 'join',
        clientId: this.self.clientId,
      })

      await this.setRtcClient()
    } catch (error) {
      console.error(error)
    }
  }

  // joinを受信した時やofferを受信したらメンバーを追加する
  async addMember(data: Member) {
    console.log('addMember', data)
    if (this.mediaStream && this.self.clientId && this.room.roomId) {
      const remoteVideoSelector = `#video-${data.clientId}`
      data.webRtc = new WebRtc(
        this.mediaStream,
        this.room.roomId,
        this.self.clientId,
        data.clientId,
        remoteVideoSelector,
        this.constraints
      )
      data.status = 'online'
      await data.webRtc.startListening()
    } else {
      console.error('no mediaStream')
    }
    const newMember = {
      [data.clientId]: data,
    }
    this.members = { ...this.members, ...newMember }
    await this.setRtcClient()
  }

  async removeMember(data: Member) {
    console.log('removeMember', this.members[data.clientId])
    if (this.members[data.clientId]) {
      this.members[data.clientId].webRtc?.disconnect()
      // delete this.members[data.clientId]
      this.members[data.clientId].status = 'offline'
    }
    await this.setRtcClient()
  }

  // シグナリングサーバーをリスンする処理
  async startListening() {
    console.log('startListening', this.self)

    // Joinに関するリスナー
    this.databaseJoinRef().on('child_added', async (snapshot) => {
      const data = snapshot.val()
      if (data === null) return
      const { clientId } = data
      if (clientId === this.self.clientId) {
        // 自分自身は無視する
        return
      }
      // 2-1. joinを受信して新メンバーの情報をローカルに登録する
      console.log('receive join', data)
      await this.addMember(data)
      // 2-2. acceptを送信する
      await this.databaseJoinRef(clientId).set({
        type: 'accept',
        clientId: this.self.clientId,
        name: this.self.name,
      })
    })
    await this.databaseJoinRef(this.self.clientId).onDisconnect().remove()

    this.databaseJoinRef(this.self.clientId).on('value', async (snapshot) => {
      const data = snapshot.val()
      if (data === null) return
      const { type, clientId } = data
      switch (type) {
        case 'accept':
          // 3-1. acceptを受信して既存メンバーの情報をローカルに登録する
          console.log('receive accept', data)
          await this.addMember(data)
          await this.members[clientId].webRtc?.offer()
          break
        default:
          break
      }
    })

    // Firebaseからメンバーが削除されたらローカルのMembersから削除
    this.databaseMembersRef().on('child_removed', async (snapshot) => {
      const data = snapshot.val()
      console.log('receive remove', data)
      if (data === null) return
      const { clientId } = data
      if (clientId === this.self.clientId) {
        // ignore self message (自分自身からのメッセージは無視する）
        return
      }
      await this.removeMember(data)
    })
    // 自分の通信が切断されたらFirebaseから自分を削除
    await this.databaseMembersRef(this.self.clientId).onDisconnect().remove()

    // ブロードキャスト通信に関するリスナー
    this.databaseBroadcastRef.on('value', async (snapshot) => {
      const data = snapshot.val()
      if (data === null) return
      const { type, clientId } = data
      if (clientId === this.self.clientId) {
        // ignore self message (自分自身からのメッセージは無視する）
        return
      }
      switch (type) {
        case 'share':
          // 画面共有
          console.log('receive share', data)
          this.share = {...this.share, clientId: data.clientId}
          await this.addShare(data)
          break
        default:
          break
      }
      console.log('databaseBroadcastRef', data)
    })

    // ダイレクト通信に関するリスナー
    const databaseDirectRef = this.databaseDirectRef(this.self.clientId)
    databaseDirectRef.on('value', (snapshot) => {
      const data = snapshot.val()
      if (data === null) return
      console.log('databaseDirectRef', data)
    })
  }

  async sendAll() {
    await this.databaseBroadcastRef.set({
      type: 'call me',
      clientId: this.self.clientId,
      message: 'I am ' + this.self.name,
    })
  }
  async sendTarget(remotePeerName: string) {
    await this.databaseDirectRef(remotePeerName).set({
      type: 'call me',
      clientId: this.self.clientId,
      message: 'I am ' + this.self.name,
    })
  }

  databaseJoinRef(path = '') {
    return getDatabase(this.room.roomId + '/_join_/' + path)
  }

  databaseMembersRef(path = '') {
    return getDatabase(this.room.roomId + '/_members_/' + path)
  }

  get databaseBroadcastRef() {
    return getDatabase(this.room.roomId + '/_broadcast_/')
  }

  databaseDirectRef(path = '') {
    return getDatabase(this.room.roomId + '/_direct_/' + path)
  }
}
