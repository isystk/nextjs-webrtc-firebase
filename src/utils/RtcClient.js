import FirebaseSignallingClient from './FirebaseSignallingClient';

const INITIAL_AUDIO_ENABLED = false;

export default class RtcClient {
  constructor(remoteVideoRef, setRtcClient) {
    const config = {
      iceServers: [{ urls: 'stun:stun.stunprotocol.org' }],
    };
    this.rtcPeerConnection = new RTCPeerConnection(config);
    this.firebaseSignallingClient = new FirebaseSignallingClient();
    this.localPeerName = '';
    this.remotePeerName = '';
    this.remoteVideoRef = remoteVideoRef;
    this._setRtcClient = setRtcClient;
    this.mediaStream = null;
  }

  get initialAudioMuted() {
    return !INITIAL_AUDIO_ENABLED;
  }

  setRtcClient() {
    this._setRtcClient(this);
  }

  async getUserMedia() {
    try {
      const constraints = { audio: true, video: true };
      this.mediaStream = await navigator.mediaDevices.getUserMedia(constraints);
    } catch (error) {
      console.error(error);
    }
  }

  async setMediaStream() {
    await this.getUserMedia();
    this.addTracks();
    this.setRtcClient();
  }

  addTracks() {
    this.addAudioTrack();
    this.addVideoTrack();
  }

  addAudioTrack() {
    this.audioTrack.enabled = INITIAL_AUDIO_ENABLED;
    this.rtcPeerConnection.addTrack(this.audioTrack, this.mediaStream);
  }

  addVideoTrack() {
    this.rtcPeerConnection.addTrack(this.videoTrack, this.mediaStream);
  }

  get audioTrack() {
    return this.mediaStream.getAudioTracks()[0];
  }

  get videoTrack() {
    return this.mediaStream.getVideoTracks()[0];
  }

  toggleAudio() {
    this.audioTrack.enabled = !this.audioTrack.enabled;
    this.setRtcClient();
  }

  async offer() {
    const sessionDescription = await this.createOffer();
    await this.setLocalDescription(sessionDescription);
    await this.sendOffer();
  }

  async createOffer() {
    try {
      return await this.rtcPeerConnection.createOffer();
    } catch (e) {
      console.error(e);
    }
  }

  async setLocalDescription(sessionDescription) {
    try {
      await this.rtcPeerConnection.setLocalDescription(sessionDescription);
    } catch (e) {
      console.error(e);
    }
  }

  async sendOffer() {
    this.firebaseSignallingClient.setPeerNames(
      this.localPeerName,
      this.remotePeerName
    );

    await this.firebaseSignallingClient.sendOffer(this.localDescription);
  }

  setOntrack() {
    this.rtcPeerConnection.ontrack = (rtcTrackEvent) => {
      if (rtcTrackEvent.track.kind !== 'video') return;

      const remoteMediaStream = rtcTrackEvent.streams[0];
      this.remoteVideoRef.current.srcObject = remoteMediaStream;
      this.setRtcClient();
    };

    this.setRtcClient();
  }

  async answer(sender, sessionDescription) {
    try {
      this.remotePeerName = sender;
      this.setOnicecandidateCallback();
      this.setOntrack();
      await this.setRemoteDescription(sessionDescription);
      const answer = await this.rtcPeerConnection.createAnswer();
      await this.rtcPeerConnection.setLocalDescription(answer);
      await this.sendAnswer();
    } catch (e) {
      console.error(e);
    }
  }

  async connect(remotePeerName) {
    this.remotePeerName = remotePeerName;
    this.setOnicecandidateCallback();
    this.setOntrack();
    await this.offer();
    this.setRtcClient();
  }

  async setRemoteDescription(sessionDescription) {
    await this.rtcPeerConnection.setRemoteDescription(sessionDescription);
  }

  async sendAnswer() {
    this.firebaseSignallingClient.setPeerNames(
      this.localPeerName,
      this.remotePeerName
    );

    await this.firebaseSignallingClient.sendAnswer(this.localDescription);
  }

  async saveReceivedSessionDescription(sessionDescription) {
    try {
      await this.setRemoteDescription(sessionDescription);
    } catch (e) {
      console.error(e);
    }
  }

  get localDescription() {
    return this.rtcPeerConnection.localDescription.toJSON();
  }

  async addIceCandidate(candidate) {
    try {
      const iceCandidate = new RTCIceCandidate(candidate);
      await this.rtcPeerConnection.addIceCandidate(iceCandidate);
    } catch (error) {
      console.error(error);
    }
  }

  setOnicecandidateCallback() {
    this.rtcPeerConnection.onicecandidate = async ({ candidate }) => {
      if (candidate) {
        await this.firebaseSignallingClient.sendCandidate(candidate.toJSON());
      }
    };
  }

  async startListening(localPeerName) {
    this.localPeerName = localPeerName;
    this.setRtcClient();
    await this.firebaseSignallingClient.remove(localPeerName);
    this.firebaseSignallingClient.database
      .ref(localPeerName)
      .on('value', async (snapshot) => {
        const data = snapshot.val();
        if (data === null) return;

        const { candidate, sender, sessionDescription, type } = data;
        switch (type) {
          case 'offer':
            await this.answer(sender, sessionDescription);
            break;
          case 'answer':
            await this.saveReceivedSessionDescription(sessionDescription);
            break;
          case 'candidate':
            await this.addIceCandidate(candidate);
            break;
          default:
            this.setRtcClient();
            break;
        }
      });
  }
}
