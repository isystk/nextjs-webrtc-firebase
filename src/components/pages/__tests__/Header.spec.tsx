import React, { useState } from 'react'
import renderer from 'react-test-renderer'
import Header from '../Header'
import '@testing-library/jest-dom/extend-expect'
import { renderHook } from '@testing-library/react-hooks'
import RtcClient from "@/utilities/RtcClient";

describe('Header', () => {
  it('Match Snapshot', () => {
    const stateRtcClient = renderHook( () => useState<RtcClient | null>(null));
    const [, setRtcClient] = stateRtcClient.result.current;
    const rtcClient = new RtcClient(setRtcClient);
    const stateMenuOpen = renderHook( () => useState<boolean>(false));
    const [isMenuOpen, setMenuOpen] = stateMenuOpen.result.current;
    if (rtcClient === null) return;
    const component = renderer.create(<Header isMenuOpen={isMenuOpen} setMenuOpen={setMenuOpen} rtcClient={rtcClient} />)
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
