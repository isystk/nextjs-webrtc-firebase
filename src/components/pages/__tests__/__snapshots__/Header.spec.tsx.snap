// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Match Snapshot 1`] = `
<header
  className="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary App-header mui-fixed MuiPaper-elevation4"
>
  <div
    className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
  >
    <div
      className="MuiGrid-root MuiGrid-container"
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </span>
      </button>
      <div
        className="App-logo"
      />
    </div>
  </div>
</header>
`;
